qa/tasks/ceph.py

```diff
@@ -416,6 +417,15 @@ def cephfs_setup(ctx, config):
 
     yield
 
+@contextlib.contextmanager
+def watchdog_setup(ctx, config):
+    cluster_name = config['cluster']
+    manager = ctx.managers[cluster_name]
+    ctx.ceph[config['cluster']].thrashers = []
```

Comment:
This thrashers ctx variable makes sense to me but this isn't the place to put it. It should be part of the ceph task somewhere during cleanup where it joins to each thrasher.

Reply:
I also don't see where you're joining with the thrasher in the current code.

You also need to update the other thrashers to use this new ctx array.

Reply:
> I also don't see where you're joining with the thrasher in the current code.

I saw  the following code for joining:
https://github.com/ceph/ceph/blob/master/qa/tasks/thrashosds.py#L212
https://github.com/ceph/ceph/blob/master/qa/tasks/mon_thrash.py#L341
https://github.com/ceph/ceph/blob/master/qa/tasks/rbd_mirror_thrash.py#L215

Do you mean iterating the thrashers[] array and do a join, as we do [here](https://github.com/ceph/ceph/blob/master/qa/tasks/mds_thrash.py#L533)?

> You also need to update the other thrashers to use this new ctx array.

I did the following changes for that.
https://github.com/ceph/ceph/pull/28378/files#diff-c3d855925f696c8644a9a7d0ce986e8bR336
https://github.com/ceph/ceph/pull/28378/files#diff-7afd9f3de4f3b5ebd97e3fb9c6200de4R206
https://github.com/ceph/ceph/pull/28378/files#diff-53114d65ead9aa4495802dfaf1dedde0R208

Please clarify.

Reply:
> > I also don't see where you're joining with the thrasher in the current code.
> 
> I saw the following code for joining:
> https://github.com/ceph/ceph/blob/master/qa/tasks/thrashosds.py#L212
> https://github.com/ceph/ceph/blob/master/qa/tasks/mon_thrash.py#L341
> https://github.com/ceph/ceph/blob/master/qa/tasks/rbd_mirror_thrash.py#L215
> 
> Do you mean iterating the thrashers[] array and do a join, as we do [here](https://github.com/ceph/ceph/blob/master/qa/tasks/mds_thrash.py#L533)?

Sorry, ignore my comment here. I see why you have the thrasher array: so the daemon watchdog can check if the thrasher has failed. So nevermind, keep the thrasher joins in each thrasher.

Reply:
Thanks, Updated the code according to that.

> This thrashers ctx variable makes sense to me but this isn't the place to put it. It should be part of the ceph task somewhere during cleanup where it joins to each thrasher.

I understand that you also concur with keeping the `thrashers` array in `watchdog_setup()` function. I will go for a teuthology run then.